march_hardware/src/Joint.cpp

```diff
@@ -77,34 +83,65 @@ void Joint::actuateRad(double target_position)
   this->imc_->actuateRad(target_position);
 }
 
-double Joint::getAngleRadAbsolute()
+void Joint::readEncoders(const ros::Duration& elapsed_time)
 {
   if (!this->hasIMotionCube())
   {
     ROS_WARN("[%s] Has no iMotionCube", this->name_.c_str());
-    return -1;
+    return;
   }
-  return this->imc_->getAngleRadAbsolute();
-}
 
-double Joint::getAngleRadIncremental()
-{
-  if (!this->hasIMotionCube())
+  if (this->receivedDataUpdate())
   {
-    ROS_WARN("[%s] Has no iMotionCube", this->name_.c_str());
-    return -1;
+    const double new_incremental_position = this->imc_->getAngleRadIncremental();
+    const double new_absolute_position = this->imc_->getAngleRadAbsolute();
+
+    // Get velocity from encoder position
+    double best_displacement;
+
+    // Take the velocity with the highest resolution.
+    if (this->imc_->getIncrementalRadPerBit() < this->imc_->getAbsoluteRadPerBit())
```

Comment:
I wouldn't evaluate this expression every time. I think you should make a boolean per joint. Can you redirect function calls in c++ like you can in Python?

Reply:
I thought this was neater, because it requires less code while the evaluation of the above expression is very close to zero.

Reply:
I thought this was neater, because it requires less code while the evaluation of the above expression is very close to zero.

Reply:
But it is a constant property of a joint that either the incremental or the absolute encoder is the most precise. It is not a big deal, but may be someone else can make the call on what should be used. I realize now you can of course not reassign a function at runtime, because it is compiled.

Reply:
You could use [`std::function`](https://en.cppreference.com/w/cpp/utility/functional) for this, but I think that would make things unnecessarily complicated. You could also just store a `bool` on which encoder has a higher resolution, saves you 2 divide operations and a float comparison. It's up to you :shrug:

Reply:
Imma leave it like this. I think it is easier to follow exactly when the condition is true this way, because you don't have to look in multiple places to find all the logic. 2 divide operations and a float comparison aren't a very big resource sink :)

Reply:
The ```bool``` suggestions seems nice. But since it's up to you I will approve the branch but not merge.